### PR TITLE
feat: validate nRF Cloud API response

### LINKED ISSUE #291: Implement nRF Cloud API response validation
These API calls could use validation:
- [ ] https://github.com/hello-nrfcloud/backend/blob/f17d6af059c525193b92b643096e0597c75f9531/lambda/resolveSingleCellGeoLocation.ts#L156
- [ ] https://github.com/hello-nrfcloud/backend/blob/f17d6af059c525193b92b643096e0597c75f9531/lambda/resolveSingleCellGeoLocation.ts#L57
- [ ] https://github.com/hello-nrfcloud/backend/blob/f17d6af059c525193b92b643096e0597c75f9531/nrfcloud/createAccountDevice.ts#L15
- [ ] https://github.com/hello-nrfcloud/backend/blob/f17d6af059c525193b92b643096e0597c75f9531/nrfcloud/devices.ts#L185
- [ ] https://github.com/hello-nrfcloud/backend/blob/f17d6af059c525193b92b643096e0597c75f9531/nrfcloud/getDeviceShadowFromnRFCloud.ts#L32

### PR DESCRIPTION
**Pull request by pudkrong**
## PR Type:
Enhancement

___
## PR Description:
This PR introduces validation for the nRF Cloud API responses. It uses the AJV (Another JSON Schema Validator) library and the TypeBox library to define and validate the schema of the responses. In case of validation errors, a ValidationError is thrown. The PR affects multiple files, mainly focusing on implementing the validation in the fetch functions and updating the function calls accordingly.

___
## PR Main Files Walkthrough:
<details> <summary>files:</summary>

`lambda/loggingFetch.ts`: Added a ValidationError class and a validate function to validate the schema of the responses. Also, introduced a new function validatedLoggingFetch which validates the response before returning it.
`lambda/resolveSingleCellGeoLocation.ts`: Replaced loggingFetch with validatedLoggingFetch and updated the function calls to handle the validation errors.
`nrfcloud/createAccountDevice.ts`: Replaced the fetch call with a call to validatedFetch and handled the validation errors.
`nrfcloud/devices.ts`: Replaced the fetch call with a call to validatedFetch and handled the validation errors.
`nrfcloud/getDeviceShadowFromnRFCloud.ts`: Replaced the fetch call with a call to validatedFetch and handled the validation errors.
`nrfcloud/validatedFetch.ts`: Introduced a new function validatedFetch which validates the response before returning it.
</details>

### CODE REVIEW

File: lambda/resolveSingleCellGeoLocation.ts [hunk 1]
```diff
@@ -41,7 +40,7 @@ const deviceFetcher = getDeviceAttributesById({ db, DevicesTableName })
 
 const { track, metrics } = metricsForComponent('singleCellGeo')
 
-const trackFetch = loggingFetch({ track, log })
+const trackFetch = validatedLoggingFetch({ track, log })
```
Comment (coderbyheart) [inline lambda/resolveSingleCellGeoLocation.ts]:
```suggestion
const trackFetch = validatedFetch(loggingFetch({ track, log }))
```

Since `loggingFetch` returns a fetch compatible function, `validatedLoggingFetch` should not care about the fact that it deals with a fetch that is augmented with logging capabilites. Instead make it accept a fetch instance as parameter. (Unix command design principle)

File: lambda/loggingFetch.ts [hunk 2]
```diff
@@ -21,3 +47,30 @@ export const loggingFetch =
 
 		return res
 	}
+
+export const validatedLoggingFetch = (
```
Comment (coderbyheart) [inline lambda/loggingFetch.ts]: Move validation aspect to a separate file. (see below)

File: lambda/resolveSingleCellGeoLocation.ts [hunk 3]
```diff
@@ -64,20 +62,17 @@ const serviceToken = once(
...
-				status: res.status,
-			})
-			throw new Error(`Acquiring service token failed: ${body} (${res.status})`)
-		}
-		const { token } = (await res.json()) as {
-			createdAt: string // e.g. '2022-12-19T19:39:02.655Z'
-			token: string // JWT
+
+		if ('error' in maybeResult) {
+			log.error('request failed', { error: maybeResult.error })
```
Comment (coderbyheart) [inline lambda/resolveSingleCellGeoLocation.ts]: Make logging request errors part of validatingFetch, so it is not forgotte, and does not have to be repeated every time.

File: lambda/resolveSingleCellGeoLocation.ts [hunk 4]
```diff
@@ -64,20 +62,17 @@ const serviceToken = once(
...
+			throw new Error(`Acquiring service token failed: ${maybeResult.error}`)
```
Comment (coderbyheart) [inline lambda/resolveSingleCellGeoLocation.ts]: **comment:** this swallows the validation errors (because the error is stringified). But OK in this case because error is logged and function is only used in lambda.

File: nrfcloud/createAccountDevice.ts [hunk 5]
```diff
@@ -12,18 +13,25 @@ export const createAccountDevice = async ({
...
-				Authorization: `Bearer ${apiKey}`,
-			},
-		})
-	).json()
+		},
+	)
+
+	if ('error' in maybeResult) {
+		console.error(`Failed to create account device:`, maybeResult.error)
+		throw new Error(`Failed to create account device`)
```
Comment (coderbyheart) [inline nrfcloud/createAccountDevice.ts]: **problem:** this swallows the error details, change API and return error

File: nrfcloud/createAccountDevice.ts [hunk 6]
```diff
@@ -12,18 +13,25 @@ export const createAccountDevice = async ({
...
-			headers: {
-				Authorization: `Bearer ${apiKey}`,
-			},
-		})
-	).json()
+		},
+	)
+
+	if ('error' in maybeResult) {
+		console.error(`Failed to create account device:`, maybeResult.error)
```
Comment (coderbyheart) [inline nrfcloud/createAccountDevice.ts]: **problem:** Do not call console.log from functions.

File: nrfcloud/devices.ts [hunk 7]
```diff
@@ -84,6 +84,10 @@ const Page = <T extends TSchema>(Item: T) =>
 	})
 const Devices = Page(Device)
 
+const ProvisionDevice = Type.Object({
+	bulkOpsRequestId: Type.String(),
+})
```
Comment (coderbyheart) [inline nrfcloud/devices.ts:92]: Add link to documentation.

File: nrfcloud/getDeviceShadowFromnRFCloud.ts [hunk 8]
```diff
@@ -1,20 +1,38 @@
+import { Type, type Static } from '@sinclair/typebox'
 import { logger } from '../lambda/util/logger.js'
-import { slashless } from '../util/slashless.js'
-import { type DeviceShadow } from './DeviceShadow.js'
+import { validatedFetch } from './validatedFetch.js'
+
+const DeviceShadow = Type.Object({
```
Comment (coderbyheart) [inline nrfcloud/getDeviceShadowFromnRFCloud.ts]: Add link to documentation

File: nrfcloud/getDeviceShadowFromnRFCloud.ts [hunk 9]
```diff
@@ -1,20 +1,38 @@
...
+	id: Type.String(),
+	state: Type.Object({
+		reported: Type.Record(Type.String(), Type.Any()),
+		version: Type.Number(),
+		metadata: Type.Record(Type.String(), Type.Any()),
```
Comment (coderbyheart) [inline nrfcloud/getDeviceShadowFromnRFCloud.ts]: Do not use any.

File: nrfcloud/validatedFetch.ts [hunk 10]
```diff
@@ -31,7 +31,9 @@ const fetchData =
 	async (...args: Parameters<typeof fetch>): ReturnType<typeof fetch> => {
 		const response = await (fetchImplementation ?? fetch)(...args)
 		if (!response.ok)
-			throw new Error(`Error fetching status: ${response.status}`)
+			throw new Error(
+				`Error fetching status: ${response.status} - ${response.statusText}`,
```
Comment (coderbyheart) [inline nrfcloud/validatedFetch.ts]: Unrelated change.

File: nrfcloud/validatedFetch.ts [hunk 11]
```diff
@@ -44,12 +46,14 @@ export const validatedFetch =
 	async <Schema extends TObject>(
 		{ resource }: { resource: string },
 		schema: Schema,
+		init?: RequestInit,
```
Comment (coderbyheart) [inline nrfcloud/validatedFetch.ts]: Unrelated change.

**Review by coderbyheart — CHANGES_REQUESTED** [review]

File: lambda/resolveSingleCellGeoLocation.ts (hunk 1 above)
Comment (pudkrong) [reply lambda/resolveSingleCellGeoLocation.ts]: I will use the existing function, [validatedFetch](https://github.com/hello-nrfcloud/backend/blob/291-implement-nrf-cloud-api-response-validation/nrfcloud/validatedFetch.ts), because

- It accepts `fetch` instance 
- However, I need to modify the function to accept 3rd parameter because the existing function will use `GET` method and `Content-Type` as `application/json` as its defaults.

File: nrfcloud/createAccountDevice.ts (hunk 5 above)
Comment (pudkrong) [reply nrfcloud/createAccountDevice.ts]: I will re-throw the error from `validatedFetch`. The error can be either `Error` object or `ValidationError`.

File: nrfcloud/getDeviceShadowFromnRFCloud.ts (hunk 9 above)
Comment (pudkrong) [reply nrfcloud/getDeviceShadowFromnRFCloud.ts]: Change to use `Type.Unknown()`

File: nrfcloud/validatedFetch.ts (hunk 11 above)
Comment (pudkrong) [reply nrfcloud/validatedFetch.ts]: This change is necessary due to the default behavior uses `GET` method and `Content-Type` as `application/json`

File: lambda/fetchDeviceShadow.ts [hunk 12]
```diff
@@ -215,7 +215,9 @@ const h = async (): Promise<void> => {
 					'shadowAge',
 					MetricUnits.Seconds,
 					Math.round(Date.now() / 1000) -
-						getShadowUpdateTime(deviceShadow.state.metadata),
+						getShadowUpdateTime(
+							deviceShadow.state.metadata as Record<string, any>,
```
Comment (pudkrong) [inline lambda/fetchDeviceShadow.ts]: I need to cast `deviceShadow.state.metadata` to `Record<string, any>` because `getShadowUpdateTime` requires

File: lambda/loggingFetch.ts [hunk 13]
```diff
@@ -4,7 +4,10 @@ import type { Logger } from '@aws-lambda-powertools/logger'
 
 export const loggingFetch =
 	({ track, log }: { track: AddMetricsFn; log: Logger }) =>
-	async (url: URL, init?: RequestInit): ReturnType<typeof fetch> => {
+	async (
+		url: URL | RequestInfo,
```
Comment (pudkrong) [inline lambda/loggingFetch.ts:8]: This is required to allow `loggingFetch` to be able to pass as `fetchImplementation`

File: lambda/validatedFetch.ts
Comment (coderbyheart) [inline lambda/validatedFetch.ts:1]: File is empty.

File: lambda/resolveSingleCellGeoLocation.ts [hunk 14]
```diff
@@ -153,9 +157,11 @@ const h = async (event: {
 				},
 			],
 		}
-		// FIXME: validate response
-		const res = await trackFetch(
-			new URL(`${slashless(apiEndpoint)}/v1/location/ground-fix`),
+
+		const vf = validatedFetch({ endpoint: apiEndpoint, apiKey }, trackFetch)
```
Comment (coderbyheart) [inline lambda/resolveSingleCellGeoLocation.ts]: locationServiceToken can be passed here.

File: lambda/fetchDeviceShadow.ts (hunk 12 above)
Comment (coderbyheart) [reply lambda/fetchDeviceShadow.ts]: Yeah, recursive typing is not really possible with TypeBox (and JSON schema)

File: nrfcloud/getDeviceShadowFromnRFCloud.ts [hunk 15]
```diff
@@ -1,20 +1,41 @@
+import { Type, type Static } from '@sinclair/typebox'
 import { logger } from '../lambda/util/logger.js'
-import { slashless } from '../util/slashless.js'
-import { type DeviceShadow } from './DeviceShadow.js'
+import { validatedFetch } from './validatedFetch.js'
+
+/**
+ * @link https://api.nrfcloud.com/v1/#tag/All-Devices/operation/ListDevices
+ */
+const DeviceShadow = Type.Object({
```
Comment (coderbyheart) [inline nrfcloud/getDeviceShadowFromnRFCloud.ts]: Should have been integrated in DeviceShadow.ts

File: lambda/loggingFetch.ts (hunk 13 above)
Comment (coderbyheart) [reply lambda/loggingFetch.ts:8]: I've addressed this in a refactor.

File: nrfcloud/validatedFetch.spec.ts [hunk 16]
```diff
@@ -50,4 +50,41 @@ describe('validatedFetch()', () => {
...
+		)
+		const schema = Type.Object({ foo: Type.Literal('bar') })
+
+		const res = await vf({ resource: 'foo' }, schema, {
+			method: 'POST',
+			headers: {
+				Authorization: 'Bearer another-key',
+				'Content-Type': 'application/octet-stream',
+			},
+		})
```
Comment (coderbyheart) [inline nrfcloud/validatedFetch.spec.ts]: If refactored this, because this mixes two parameters to configure the fetch behaviour. Those two (the first and the third) are also split around the schema.

File: nrfcloud/validatedFetch.ts (hunk 11 above)
Comment (coderbyheart) [reply nrfcloud/validatedFetch.ts]: Changed in the refactor.

**Review by coderbyheart — APPROVED** [review]
